pydm_alarm_plugin/alarm_plugin.py

```diff
@@ -0,0 +1,113 @@
+import logging
+import os
+from kafka.consumer.fetcher import ConsumerRecord
+from pydm.data_plugins.plugin import PyDMPlugin, PyDMConnection
+from pydm.utilities import remove_protocol
+from pydm.widgets.base import PyDMWidget
+from pydm.widgets.channel import PyDMChannel
+from qtpy.QtCore import QObject, QThread
+from slam.alarm_item import AlarmSeverity
+from slam.kafka_reader import KafkaReader
+from typing import Optional
+
+
+logger = logging.getLogger(__name__)
+kafka_logger = logging.getLogger('kafka')
+kafka_logger.setLevel('WARNING')  # Anything less results in too much noise
+
+
+class Connection(PyDMConnection):
+    """
+    A PyDMConnection for sending alarm severity signals to all listeners which have been registered with it.
+    """
+    def __init__(self, channel: PyDMChannel, address: str, protocol: Optional[str] = None,
+                 parent: Optional[QObject] = None):
+        super().__init__(channel, address, protocol, parent)
+        self.add_listener(channel)
+
+    def send_alarm_data(self, severity: AlarmSeverity):
+        """
+        Sends the updated alarm severity to all listeners of this connection.
+
+        Parameters
+        ----------
+        severity : AlarmSeverity
+            The updated severity which will be emitted as the new_severity_signal
+        """
+        if severity is AlarmSeverity.OK:
+            self.new_severity_signal.emit(PyDMWidget.ALARM_NONE)
+        elif severity is AlarmSeverity.MINOR:
+            self.new_severity_signal.emit(PyDMWidget.ALARM_MINOR)
+        elif severity is AlarmSeverity.MAJOR:
+            self.new_severity_signal.emit(PyDMWidget.ALARM_MAJOR)
+        elif severity is AlarmSeverity.INVALID:
+            self.new_severity_signal.emit(PyDMWidget.ALARM_INVALID)
+        elif severity is AlarmSeverity.UNDEFINED:
+            self.new_severity_signal.emit(PyDMWidget.ALARM_DISCONNECTED)
+
+
+class AlarmPlugin(PyDMPlugin):
+    """
+    Manages the data flow between the kafka queue for alarm data and the PyDM display widgets. Currently this is
+    read-only as no write actions can be taken from PyDM widgets for now.
+    """
+    protocol = "nalms"
+    connection_class = Connection
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        kafka_topic = os.getenv('PYDM_KAFKA_ALARM_TOPIC')
```

Comment:
Does this implementation mean that each pydm process with `nalms://` channels will hear all of the updates from kafka, and only use the ones that are relevant to it?

Is there any performance hit here for the initial startup where we need to collect and cache all of the existing alarm data? Or not, maybe the data set isn't huge.

Reply:
Yes that is correct, there is no way to ask for specific keys from the queue. Good question about performance, there will be an initial hit for reading from the queue, but I think it will be negligible for the size of the average data set. I am collecting some performance numbers for different amounts of alarms to be sure.

Reply:
On a performance profile the kafka related calls are below the most costly ones. For some actual numbers (that are specific to the box it was run on of course - shared machine with moderate load) it can go through 1000 records in ~150 milliseconds. 10,000 records in ~800 ms, and on the higher side 90,000 in about 6.5 seconds.

1000 records should be above the upper end of our initial deployment (a few hundred). But even in the more extreme case of ~100,000 the display will load quickly and remain responsive while processing the records since that happens in a separate thread. It would take the full time to ensure all the alarm colors are correct and not ideal if any costly processing for the display is going on that it would be competing with.

There is a way this could be made more performant I was thinking of, where on startup instead of reading from the beginning of the kafka queue it just listens for new messages only. And as connections to the `nalms://` protocol are made, in order to give it the current value of the alarm it reads from an elasticsearch index instead (already part of NALMS) if we haven't yet seen that alarm in kafka, and then any further updates come from kafka. This way the process the entire queue step could be skipped. This would be straightforward to implement I believe, would want to run numbers on that as well to ensure it is better.

I think at the moment I'd lean towards leaving this as it currently is for small data sets, and add an issue for follow up for better performance.

Reply:
I agree, if the potential performance issues don't appear until ~1e6 records then there is no sense in using dev time to prematurely optimize.

The potential issue you've laid out here (some random channels getting first update at up 6.5s later instead of immediately based on luck at the order we get them) doesn't even seem that bad.

I like the path you've laid out to potential startup speedups and it may be worth doing if someone manages to configure 100,000 alarms somehow. I suspect we won't even reach 10,000.

I'm wondering if at 100,000 if we'd also start to see issues with too many consumers asking for too much data all at once. Imagine the contrived case where every pydm display has a `nalms://` channel and we hammer `kafka` for 100,000 records on concurrent screen opens from multiple endstations.

But that's all imaginary scaling and I'm :+1: on keeping it as is here.